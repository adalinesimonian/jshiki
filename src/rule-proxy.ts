import RuleTree, { AccessPathArray, AccessPathPart } from './rule-tree'

function hasAccess(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  propertyScope: AccessPathArray
) {
  const match = ruleTree.match(propertyScope)
  return explicitAllow
    ? match === true || (typeof match === 'string' && match === 'allow')
    : !(typeof match === 'string' && match === 'block')
}

function getProxyHandler<T extends object>(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  scope: AccessPathArray = []
): ProxyHandler<T> {
  const cache = new Map<AccessPathPart, any>()

  return {
    get(target: T, property: AccessPathPart) {
      if (cache.has(property)) {
        return cache.get(property)
      }

      const propertyScope = scope.concat(property)

      if (!hasAccess(ruleTree, explicitAllow, propertyScope)) {
        cache.set(property, undefined)
        return undefined
      }

      const propValue = (target as any)[property]
      const returnValue =
        (propValue && typeof propValue === 'object') ||
        typeof propValue === 'function'
          ? new Proxy(
              (target as any)[property],
              getProxyHandler(ruleTree, explicitAllow, propertyScope)
            )
          : propValue
      cache.set(property, returnValue)
      return returnValue
    },

    set(_target: T, _property: AccessPathPart, value: any) {
      return value
    },
  }
}

export default function getRuleProxy<T extends object>(
  target: T,
  ruleTree: RuleTree,
  explicitAllow?: boolean
): T {
  return (target && typeof target === 'object') || typeof target === 'function'
    ? new Proxy(target, getProxyHandler(ruleTree, Boolean(explicitAllow)))
    : target
}
