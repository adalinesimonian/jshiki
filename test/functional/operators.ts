import { evaluate } from '../../src'

describe('Operators', () => {
  describe('Unary Operators', () => {
    describe('+ (Unary plus)', () => {
      it('should convert to numbers', () => {
        expect(evaluate('+5')).toBe(5)
        expect(evaluate("+'5'")).toBe(5)
        expect(evaluate('+true')).toBe(1)
        expect(evaluate('+false')).toBe(0)
        expect(evaluate('+null')).toBe(0)
      })
    })

    describe('- (Unary negation)', () => {
      it('should negate numbers', () => {
        expect(evaluate('-5')).toBe(-5)
      })
    })

    describe('! (Logical NOT)', () => {
      it('should negate expressions', () => {
        expect(evaluate('!true')).toBe(false)
        expect(evaluate('!false')).toBe(true)
        expect(evaluate('!1')).toBe(false)
        expect(evaluate('!0')).toBe(true)
        expect(evaluate('!!true')).toBe(true)
        expect(evaluate('!!false')).toBe(false)
        expect(evaluate('!!1')).toBe(true)
        expect(evaluate('!!0')).toBe(false)
      })
    })

    describe('~ (Bitwise NOT)', () => {
      it('should invert numbers', () => {
        expect(evaluate('~5')).toBe(-6)
        expect(evaluate('~-15')).toBe(14)
      })
    })
  })

  describe('Binary Operators', () => {
    describe('+ (Addition)', () => {
      it('should add integers', () => {
        expect(evaluate('5 + 3')).toBe(8)
      })

      it('should add floating-point numbers', () => {
        expect(evaluate('5.2 + 3.8')).toBe(9)
      })

      it('should concatenate strings', () => {
        expect(evaluate("'x' + 'y'")).toBe('xy')
      })
    })

    describe('- (Subtraction)', () => {
      it('should subtract integers', () => {
        expect(evaluate('5 - 3')).toBe(2)
      })

      it('should subtract floating-point numbers', () => {
        expect(evaluate('5.5 - 3.2')).toBe(2.3)
      })
    })

    describe('* (Multiplication)', () => {
      it('should multiply integers', () => {
        expect(evaluate('5 * 3')).toBe(15)
      })

      it('should multiply floating-point numbers', () => {
        expect(evaluate('5.4 * 3.6')).toBe(19.44)
      })
    })

    describe('** (Exponential)', () => {
      it('should raise first operand to power of second operand', () => {
        expect(evaluate('5 ** 3')).toBe(125)
      })

      it('should exponentiate floating-point numbers', () => {
        expect(evaluate('2.1 ** 3.5')).toBe(2.1 ** 3.5)
      })
    })

    describe('/ (Division)', () => {
      it('should divide integers', () => {
        expect(evaluate('8 / 4')).toBe(2)
        expect(evaluate('5 / 3')).toBe(5 / 3)
      })

      it('should divide floating-point numbers', () => {
        expect(evaluate('8.2 / 4.2')).toBe(8.2 / 4.2)
        expect(evaluate('5.5 / 3.5')).toBe(5.5 / 3.5)
      })
    })

    describe('% (Modulo)', () => {
      it('should perform modulo on integers', () => {
        expect(evaluate('8 % 4')).toBe(0)
        expect(evaluate('5 % 3')).toBe(2)
      })
    })

    describe('< (Less than)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 < 4')).toBe(false)
        expect(evaluate('3 < 5')).toBe(true)
        expect(evaluate('5 < 5')).toBe(false)
      })
    })

    describe('> (Greater than)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 > 4')).toBe(true)
        expect(evaluate('3 > 5')).toBe(false)
        expect(evaluate('5 > 5')).toBe(false)
      })
    })

    describe('<= (Less than or equal to)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 <= 4')).toBe(false)
        expect(evaluate('3 <= 5')).toBe(true)
        expect(evaluate('5 <= 5')).toBe(true)
      })
    })

    describe('>= (Greater than or equal to)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 >= 4')).toBe(true)
        expect(evaluate('3 >= 5')).toBe(false)
        expect(evaluate('5 >= 5')).toBe(true)
      })
    })

    /* eslint-disable eqeqeq */
    describe('== (Equality)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 == 4')).toBe(false)
        expect(evaluate('3 == 5')).toBe(false)
        expect(evaluate('5 == 5')).toBe(true)
      })

      it('should compare strings', () => {
        expect(evaluate("'1' == '1'")).toBe(true)
        expect(evaluate("'1' == '2'")).toBe(false)
      })

      it('should compare objects', () => {
        expect(evaluate("{ x: '1' } == { x: '1' }")).toBe(false)
      })

      it('should compare null and undefined', () => {
        expect(evaluate('null == undefined')).toBe(true)
      })
    })

    describe('!= (Inequality)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 != 4')).toBe(true)
        expect(evaluate('3 != 5')).toBe(true)
        expect(evaluate('5 != 5')).toBe(false)
      })

      it('should compare strings', () => {
        expect(evaluate("'1' != '1'")).toBe(false)
        expect(evaluate("'1' != '2'")).toBe(true)
      })

      it('should compare objects', () => {
        expect(evaluate("{ x: '1' } != { x: '1' }")).toBe(true)
      })

      it('should compare null and undefined', () => {
        expect(evaluate('undefined != null')).toBe(false)
      })
    })
    /* eslint-enable eqeqeq */

    describe('=== (Identity)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 === 4')).toBe(false)
        expect(evaluate('3 === 5')).toBe(false)
        expect(evaluate('5 === 5')).toBe(true)
      })

      it('should compare strings', () => {
        expect(evaluate("'1' === '1'")).toBe(true)
        expect(evaluate("'1' === '2'")).toBe(false)
      })

      it('should compare objects', () => {
        expect(evaluate("{ x: '1' } === { x: '1' }")).toBe(false)
      })

      it('should compare null and undefined', () => {
        expect(evaluate('undefined === null')).toBe(false)
      })
    })

    describe('!== (Non-identity)', () => {
      it('should compare numbers', () => {
        expect(evaluate('8 !== 4')).toBe(true)
        expect(evaluate('3 !== 5')).toBe(true)
        expect(evaluate('5 !== 5')).toBe(false)
      })

      it('should compare strings', () => {
        expect(evaluate("'1' !== '1'")).toBe(false)
        expect(evaluate("'1' !== '2'")).toBe(true)
      })

      it('should compare objects', () => {
        expect(evaluate("{ x: '1' } !== { x: '1' }")).toBe(true)
      })

      it('should compare null and undefined', () => {
        expect(evaluate('undefined !== null')).toBe(true)
      })
    })

    describe('&& (Logical AND)', () => {
      it('should compare booleans', () => {
        expect(evaluate('false && false')).toBe(false)
        expect(evaluate('false && true')).toBe(false)
        expect(evaluate('true && false')).toBe(false)
        expect(evaluate('true && true')).toBe(true)
      })

      it('should compare numbers', () => {
        expect(evaluate('0 && 0')).toBe(0)
        expect(evaluate('0 && 1')).toBe(0)
        expect(evaluate('1 && 0')).toBe(0)
        expect(evaluate('1 && 1')).toBe(1)
      })

      it('should compare strings', () => {
        expect(evaluate("'' && ''")).toBe('')
        expect(evaluate("'' && '1'")).toBe('')
        expect(evaluate("'1' && ''")).toBe('')
        expect(evaluate("'1' && '1'")).toBe('1')
      })

      it('should compare null and undefined', () => {
        expect(evaluate('null && undefined')).toBe(null)
        expect(evaluate('undefined && null')).toBe(undefined)
        expect(evaluate("null && 'x'")).toBe(null)
        expect(evaluate("undefined && 'x'")).toBe(undefined)
        expect(evaluate("'x' && null")).toBe(null)
        expect(evaluate("'x' && undefined")).toBe(undefined)
      })
    })

    describe('|| (Logical OR)', () => {
      it('should compare booleans', () => {
        expect(evaluate('false || false')).toBe(false)
        expect(evaluate('false || true')).toBe(true)
        expect(evaluate('true || false')).toBe(true)
        expect(evaluate('true || true')).toBe(true)
      })

      it('should compare numbers', () => {
        expect(evaluate('0 || 0')).toBe(0)
        expect(evaluate('0 || 1')).toBe(1)
        expect(evaluate('1 || 0')).toBe(1)
        expect(evaluate('1 || 1')).toBe(1)
      })

      it('should compare strings', () => {
        expect(evaluate("'' || ''")).toBe('')
        expect(evaluate("'' || '1'")).toBe('1')
        expect(evaluate("'1' || ''")).toBe('1')
        expect(evaluate("'1' || '1'")).toBe('1')
      })

      it('should compare null and undefined', () => {
        expect(evaluate('null || undefined')).toBe(undefined)
        expect(evaluate('undefined || null')).toBe(null)
        expect(evaluate("null || 'x'")).toBe('x')
        expect(evaluate("undefined || 'x'")).toBe('x')
        expect(evaluate("'x' || null")).toBe('x')
        expect(evaluate("'x' || undefined")).toBe('x')
      })
    })

    describe('& (Bitwise AND)', () => {
      it('should bitwise AND integers', () => {
        expect(evaluate('5 & 3')).toBe(1)
      })
    })

    describe('| (Bitwise OR)', () => {
      it('should bitwise OR integers', () => {
        expect(evaluate('5 | 3')).toBe(7)
      })
    })

    describe('^ (Bitwise XOR)', () => {
      it('should bitwise XOR integers', () => {
        expect(evaluate('5 ^ 3')).toBe(6)
      })
    })

    describe('<< (Bitwise Left Shift)', () => {
      it('should shift integer to the left', () => {
        expect(evaluate('5 << 1')).toBe(10)
      })
    })

    describe('>> (Bitwise Right Shift)', () => {
      it('should shift integer to the right', () => {
        expect(evaluate('5 >> 1')).toBe(2)
      })
    })

    describe('>>> (Bitwise Unsigned Right Shift)', () => {
      it('should unsigned shift integer to the right', () => {
        expect(evaluate('5 >>> 1')).toBe(2)
      })
    })

    describe('= (Assignment)', () => {
      it('should throw an error', () => {
        expect(() => evaluate('x = 5')).toThrowErrorMatchingSnapshot()
      })
    })
  })

  describe('Ordering', () => {
    it('should follow the order of operations', () => {
      expect(evaluate('24 / 6 * 2 + 4 / 2')).toBe(10)
      expect(evaluate('(24 / (6 * 2) + 4) / 2')).toBe(3)
    })
  })
})
