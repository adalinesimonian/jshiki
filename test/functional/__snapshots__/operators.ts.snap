// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operators (async) ?: (Conditional/Ternary) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators != (Inequality) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators !== (Non-identity) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators % (Modulo) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators & (Bitwise AND) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators * (Multiplication) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators ** (Exponential) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators + (Addition) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators ++, -- (Increment, Decrement) should throw an error 1`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators ++, -- (Increment, Decrement) should throw an error 2`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators ++, -- (Increment, Decrement) should throw an error 3`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators ++, -- (Increment, Decrement) should throw an error 4`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators ++, -- (Increment, Decrement) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:15)"`;

exports[`Operators (async) Binary Operators ++, -- (Increment, Decrement) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:15)"`;

exports[`Operators (async) Binary Operators / (Division) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators < (Less than) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators << (Bitwise Left Shift) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators <= (Less than or equal to) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 1`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 2`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 3`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 4`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 5`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 6`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 7`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 8`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 9`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 10`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 11`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 12`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 13`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 14`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 15`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 16`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 17`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 18`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 19`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw an error 20`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 3`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 4`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 5`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 6`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 7`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 8`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 9`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 10`] = `"Unexpected token (1:19)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 11`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 12`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 13`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 14`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 15`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 16`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 17`] = `"Unexpected token (1:15)"`;

exports[`Operators (async) Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 18`] = `"Unexpected token (1:15)"`;

exports[`Operators (async) Binary Operators == (Equality) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators === (Identity) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators > (Greater than) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators >= (Greater than or equal to) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators >> (Bitwise Right Shift) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators >>> (Bitwise Unsigned Right Shift) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:18)"`;

exports[`Operators (async) Binary Operators ^ (Bitwise XOR) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators | (Bitwise OR) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators - (Subtraction) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:16)"`;

exports[`Operators (async) Binary Operators in (Property Membership) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Binary Operators in (Property Membership) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:13)"`;

exports[`Operators (async) Binary Operators instanceof (Prototype Check) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:25)"`;

exports[`Operators (async) Binary Operators instanceof (Prototype Check) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:13)"`;

exports[`Operators (async) Logical Operators && (Logical AND) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Logical Operators ?? (Null coalescing) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Logical Operators || (Logical OR) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:17)"`;

exports[`Operators (async) Unary Operators Unsupported operators should throw an error 1`] = `"Unsupported unary operator: delete"`;

exports[`Operators (async) should throw when the expression contains an invalid operator 1`] = `"Unexpected token (1:21)"`;

exports[`Operators (async) should throw when the expression contains an invalid operator 2`] = `"Unexpected token (1:15)"`;

exports[`Operators ?: (Conditional/Ternary) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators != (Inequality) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators !== (Non-identity) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators % (Modulo) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators & (Bitwise AND) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators * (Multiplication) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators ** (Exponential) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators + (Addition) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators ++, -- (Increment, Decrement) should throw an error 1`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators ++, -- (Increment, Decrement) should throw an error 2`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators ++, -- (Increment, Decrement) should throw an error 3`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators ++, -- (Increment, Decrement) should throw an error 4`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators ++, -- (Increment, Decrement) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:9)"`;

exports[`Operators Binary Operators ++, -- (Increment, Decrement) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:9)"`;

exports[`Operators Binary Operators / (Division) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators < (Less than) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators << (Bitwise Left Shift) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators <= (Less than or equal to) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 1`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 2`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 3`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 4`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 5`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 6`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 7`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 8`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 9`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 10`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 11`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 12`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 13`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 14`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 15`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 16`] = `"Unsupported node type: AssignmentExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 17`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 18`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 19`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw an error 20`] = `"Unsupported node type: UpdateExpression"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 3`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 4`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 5`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 6`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 7`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 8`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 9`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 10`] = `"Unexpected token (1:13)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 11`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 12`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 13`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 14`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 15`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 16`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 17`] = `"Unexpected token (1:9)"`;

exports[`Operators Binary Operators =, +=, -=, ... (Assignment) should throw when the expression is missing an operand 18`] = `"Unexpected token (1:9)"`;

exports[`Operators Binary Operators == (Equality) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators === (Identity) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators > (Greater than) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators >= (Greater than or equal to) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators >> (Bitwise Right Shift) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators >>> (Bitwise Unsigned Right Shift) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:12)"`;

exports[`Operators Binary Operators ^ (Bitwise XOR) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators | (Bitwise OR) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators - (Subtraction) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:10)"`;

exports[`Operators Binary Operators in (Property Membership) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Binary Operators in (Property Membership) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:7)"`;

exports[`Operators Binary Operators instanceof (Prototype Check) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:19)"`;

exports[`Operators Binary Operators instanceof (Prototype Check) should throw when the expression is missing an operand 2`] = `"Unexpected token (1:7)"`;

exports[`Operators Logical Operators && (Logical AND) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Logical Operators ?? (Null coalescing) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Logical Operators || (Logical OR) should throw when the expression is missing an operand 1`] = `"Unexpected token (1:11)"`;

exports[`Operators Unary Operators Unsupported operators should throw an error 1`] = `"Unsupported unary operator: delete"`;

exports[`Operators should throw when the expression contains an invalid operator 1`] = `"Unexpected token (1:15)"`;

exports[`Operators should throw when the expression contains an invalid operator 2`] = `"Unexpected token (1:9)"`;
